File: Searching and Sorting/Binary Search/BinarySearch.py
#For binary search to execute, list/array to be searched, must be a sorted sequence.

#Call 'main()' in the terminal/console to run Binary Search on desired sorted sequence.

#This algorithm returns the presence and position of number to be searched, 
#in the sorted sequence.

def BinarySearch(arr, lb, ub, num):
    flag, mid = 0, 0
    while lb <= ub:
        mid = (lb+ub)/2
        if arr[mid] == num:
            flag = 1
            break
        elif arr[mid] < num:
            lb = mid + 1
            BinarySearch(arr, lb, ub, num)
        else:
            ub = mid - 1
            BinarySearch(arr, lb, ub, num)
    return flag, mid
            
def main():
    arr = list()
    print("Enter the elements of array in increasing/deceasing order.\
          Type 'x' (along with inverted commas) for termination")
    while True:
        el = input()
        if el is 'x' or el is 'X':
            break
        else:
            arr.append(int(el))
    print("Enter the number to be searched")
    num = int(input())
    ub = len(arr)
    flag, pos = BinarySearch(arr, 0, ub, num)
    if flag == 1:
        print("Number is present in the array at {} postion".format(pos+1))
    else:
        print("Number is not present in the array")
    return 0

if __name__ == "__main__":
    main()

File: Searching and Sorting/Binary Search/C++/Binary Search.cpp
// Coding Binary Search Algorithm through Functions Topic Approach 
#include<iostream>
using namespace std ; 
//Function that implements the binary search algorithm
int BinarySearch(int arr[] ,int n , int val )
{
	int start=0 ; 
	int end=n-1 ;
	int mid ; 
	  

	while(start<=end)
	{
		mid=(start+end)/2 ;
		if(val==arr[mid])
		{
			return mid ; 
		}
		else if(val > arr[mid])
		{
			start=mid+1 ; 

		}
		else
		{
			end=mid-1 ; 
		}

	}
	return -1 ; 

}
//Driver code for the function
int main()
{
	int arr[50],n ;
	int result ;  
	cout << "Enter the no. of elements do u want to enter" ;
	cin >> n ; 

	cout << "Enter the elements that you want to insert in the Array " << endl ; 
	for(int i=0 ; i<n ; i++)
	{
		cin >> arr[i] ; 
	}

	cout << "Enter the element that you want to search for in the Array" << endl ; 
	int val ; 
	cin >> val ; 

	result= BinarySearch(arr,n,val) ; 
	cout << result << endl ; 

}
